assets/src/components/map.tsx

```diff
@@ -217,41 +223,108 @@ export const strokeOptions = ({ color }: Shape): object =>
-    point.lat,
-    point.lon,
-  ])
+const StopMarker = React.memo(
+  ({
+    stop,
+    iconSize,
+  }: {
+    stop: Stop
+    iconSize: StationIconSize
+  }): JSX.Element => {
+    if (stop.locationType === LocationType.Station) {
+      return <StationMarker station={stop} iconSize={iconSize} />
+    }
+
+    return (
+      <CircleMarker
+        className="m-vehicle-map__stop"
+        center={[stop.lat, stop.lon]}
+        radius={3}
+      >
+        <Popup className="m-vehicle-map__stop-tooltip">
+          {stop.name}
+          {inTestGroup(MAP_BETA_GROUP_NAME) && (
+            <StreetViewButton
+              latitude={stop.lat}
+              longitude={stop.lon}
+            ></StreetViewButton>
+          )}
+        </Popup>
+      </CircleMarker>
+    )
+  }
+)
+
+const StationMarker = ({
+  station,
+  iconSize,
+}: {
+  station: Stop
+  iconSize: StationIconSize
+}) => {
+  const iconSizeLength = iconSize === StationIconSize.small ? 12 : 16
+
+  return (
+    <Marker
+      position={[station.lat, station.lon]}
+      icon={stationLeafletIcon({ size: iconSizeLength })}
+    >
+      <Tooltip
+        className="m-vehicle-map__station-tooltip"
+        direction={"top"}
+        offset={[0, -(iconSizeLength / 2 + 8)]}
+      >
+        {station.name}
+      </Tooltip>
+    </Marker>
+  )
+}
+
+enum StationIconSize {
+  small,
+  large,
+}
+
+const RouteStops = ({
+  stops,
+  iconSize,
+}: {
+  stops: Stop[]
+  iconSize: StationIconSize
+}): JSX.Element => {
+  const seenStopIds = new Set<StopId>()
+  // Keep the first occurance of each stop when there are duplicates
+  const uniqueStops: Stop[] = stops.flatMap((stop) => {
+    if (!seenStopIds.has(stop.id)) {
+      seenStopIds.add(stop.id)
+      return [stop]
+    }
+    return []
+  })
 
   return (
     <>
-      <Polyline
-        className="m-vehicle-map__route-shape"
-        positions={positions}
-        {...strokeOptions(shape)}
-      />
-      {(shape.stops || []).map((stop) => (
-        <CircleMarker
-          key={stop.id}
-          className="m-vehicle-map__stop"
-          center={[stop.lat, stop.lon]}
-          radius={3}
-        >
-          <Popup className="m-vehicle-map__stop-tooltip">
-            {stop.name}
-            {inTestGroup(MAP_BETA_GROUP_NAME) && (
-              <StreetViewButton
-                latitude={stop.lat}
-                longitude={stop.lon}
-              ></StreetViewButton>
-            )}
-          </Popup>
-        </CircleMarker>
+      {uniqueStops.map((stop) => (
+        <StopMarker key={stop.id} stop={stop} iconSize={iconSize} />
       ))}
     </>
   )
 }
 
+const RouteShape = React.memo(({ shape }: { shape: Shape }) => {
+  const positions: LatLngExpression[] = shape.points.map((point) => [
+    point.lat,
+    point.lon,
+  ])
+  return (
+    <Polyline
+      className="m-vehicle-map__route-shape"
+      positions={positions}
+      {...strokeOptions(shape)}
+    />
+  )
+})
```

Comment:
**praise:** I like separating out the shape and stops components, makes things a little easier to read in my opinion.